h2o-hive/src/main/java/water/hive/HiveFrameSaverImpl.java

```diff
@@ -0,0 +1,167 @@
+            String res = fs.getUri().toString() + "/" + configuredTmpPath;
+            LOG.info("Adding file system prefix to relative tmp_path " + res);
+            return res;
+        } else {
+            return configuredTmpPath;
+        }
+    }
+    
+    private String getRandomFileName() {
+        return "h2o_save_to_hive_" + UUID.randomUUID().toString() + ".csv";
+    }
+
+    private void safelyRemoveCsv(String csvPath) {
+        try {
+            Persist p = H2O.getPM().getPersistForURI(URI.create(csvPath));
+            if (p.exists(csvPath)) {
+                p.delete(csvPath);
+            } else {
+                LOG.debug("CSV file moved by Hive, doing nothing.");
+            }
+        } catch (Exception e) {
+            LOG.error("Failed cleaning up CSV file.", e);
+        }
+    }
+
+    private void writeFrameToHdfs(Frame f, String csvPath) throws IOException {
+        Persist p = H2O.getPM().getPersistForURI(URI.create(csvPath));
+        OutputStream os = p.create(csvPath, false);
+        InputStream is = f.toCSV(new Frame.CSVStreamParams().setHeaders(false));
+        IOUtils.copyStream(is, os);
+        os.close();
+    }
+    
+    private void loadCsvIntoTable(String url, String table, Frame frame, String csvPath) throws IOException {
+        try (Connection conn = SQLManager.getConnectionSafe(url, null, null)) {
+            if (!doesTableExist(conn, table)) {
+                createTable(conn, table, frame);
+            } else {
+                throw new IllegalArgumentException("Table " + table + " already exists.");
+            }
+            executeDataLoad(conn, table, csvPath);
+        } catch (SQLException e) {
+            throw new IOException("Failed to load data into Hive table.", e);
+        }
+    }
+
+    private boolean doesTableExist(Connection conn, String table) {
+        try (Statement stmt = conn.createStatement()) {
+            try (ResultSet rs = stmt.executeQuery(String.format(SQL_DESCRIBE_TABLE, table))) {
+                return rs.next();
+            }
+        } catch (SQLException e) {
+            return false;
+        }
+    }
+
+    private void createTable(Connection conn, String table, Frame frame) throws SQLException {
+        try (Statement stmt = conn.createStatement()) {
+            String createQuery = makeCreateTableStatement(table, frame);
+            LOG.info("Creating Hive table " + table + " with SQL: " + createQuery);
+            stmt.execute(createQuery);
+        }
+    }
+
+    private String makeCreateTableStatement(String table, Frame frame) {
+        StringBuilder sb = new StringBuilder();
+        sb.append("CREATE TABLE ").append(table).append(" (");
+        for (int i = 0; i < frame.numCols(); i++) {
+            if (i > 0) sb.append(",\n");
+            sb.append(frame.name(i)).append(" ").append(sqlType(frame.vec(i).get_type()));
+        }
+        sb.append(") ROW FORMAT SERDE 'org.apache.hadoop.hive.serde2.OpenCSVSerde'\n" +
```

Comment:
This is ok, because `https://stackoverflow.com/questions/1532461/stringbuilder-vs-string-concatenation-in-tostring-in-java` - I'll take this information from SO as a fact.